### PR TITLE
[Event Hubs] Optimize partition resolver compute hash

### PR DESCRIPTION
**Pull request by danielmarbach**
Relates to https://github.com/Azure/azure-sdk-for-net/issues/27030

[Benchmark](https://github.com/danielmarbach/MicroBenchmarks/commit/d1c45dd7fe90326b7c49513c938fdcf078990c1e) uses the same string of various sizes (see size column in the screenshot) as a partition key.

## With SkipLocalsInit

image

## Without SkipLocalsInit

image

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (danielmarbach) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:9]: These methods are actually available on NETStandard 2.1 and NetCoreApp. Should we ifdef here? If so any advises what types of conditions we would normally use?

Example

https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.getbytes?view=net-6.0#system-text-encoding-getbytes(system-readonlyspan((system-char))-system-span((system-byte)))

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs
Comment (danielmarbach) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs:1]: The original file has https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/SkipLocalsInitAttribute.cs#L1 is it a problem if we change it? I'd say not because in order to shim they are expected to be copied.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (danielmarbach) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:109]: I used 256 as a max size similar to what Json does here

https://github.com/dotnet/runtime/blob/ab02bf6745a80d991ccfbdac0f4e029b9d16f14b/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L74-L76

since it is hard to come up with a good threshold

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (danielmarbach) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:116]: I assumed we don't bother clearing the buffer here

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs:1]: Given SkipLocalsInit becomes only relevant for slightly larger stack alloc byte arrays we might also decide to get rid of it depending on the discussion (or consider it slightly too esoteric)

**Comment by danielmarbach** [comment]
@jsquire I think these are great results. I'm sure we could squeeze out more by actually also doing something with the hash algorithm but I felt this might be too complex for now for me to dive into and like you mentioned we would need to be super extra careful not not break anything. With these changes we can get nice gains without really touching the hashing function that much.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:9]: Well the benefit of ifdef would be that instead of using the local implementation the code path would be using the framework built-in versions and hence automatically benefit from any optimizations being done there

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs:1]: I'm totally ok with that

**Comment by danielmarbach** [comment]
SkipLocalsInit is removed. Last oustanding question is whether to ifdef or not. Otherwise I guess this is more or less ready to rumble.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:9]: The question that I'm considering is whether those potential optimizations are likely to be impactful enough that it warrants the lack of testability that we'd have and, to a lesser extend, warrants future maintainers (including future me) having to reason about what's being called.   If we continue to accept that the big win is that you got us to non-allocating, would saving a couple of ns make an impact?   On the other hand, wrapping the entire class in an `#ifdef` with a comment explaining things isn't a ton of cognitive burden or complexity.

I remain pretty neutral, but I'd slant towards leaving it without the `#ifdef`, potentially with a comment to reconsider if we start to explicitly support `netstandard2.1` as a target.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:11]:
```suggestion
        public static unsafe int GetBytes(this Encoding encoding, 
                                           ReadOnlySpan<char> src, 
                                           Span<byte> dest)
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:36]:
```suggestion
        public static unsafe int GetByteCount(this Encoding encoding, 
                                              ReadOnlySpan<char> src)
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:7]: If you'd be so kind, doc comments would be appreciated on the class and all methods.

**Review by jsquire — COMMENTED** [review]
Since we think we're just about ready, I gave it the nitpick formatting pass for consistency with other Event Hubs assets.


By chance, did you run any tests comparing the output of the old and new implementation across a large set of random strings?    If not, I'll do that before we merge for confidence that we didn't accidentally change behavior.

**Comment by danielmarbach** [comment]
@jsquire 

> By chance, did you run any tests comparing the output of the old and new implementation across a large set of random strings? If not, I'll do that before we merge for confidence that we didn't accidentally change behavior.

I did not yet. Normally, what I would do is write an explicit test that copies the old implementation local into the test and then exercise the new implementation together with the old and verify that for the same input the same output occurs on both sides. If you are OK with that approach, I'm happy to push that.

**Comment by jsquire** [comment]
> I did not yet. Normally, what I would do is write an explicit test that copies the old implementation local into the test and then exercise the new implementation together with the old and verify that for the same input the same output occurs on both sides. If you are OK with that approach, I'm happy to push that.

I wasn't even thinking that level of rigor.   I was going to copy the old + new into LinqPad and just run in a giant loop with a few tens of thousands of random inputs.

**Comment by jsquire** [comment]
We ended up with inconsistent results compared to the original implementation.   I traced this back to `BitConverter` reading with the correct endian-ness  for the host where the new approach is using `BinaryPrimitives` and assuming big endian reads.    

I was able to work around this by aliasing the `BinaryPrimitives` reads:

```csharp
private delegate uint ReadUInt32(ReadOnlySpan<byte> bytes);

private static ReadUInt32 ReadUInt32CorrectEndian = BitConverter.IsLittleEndian
    ? BinaryPrimitives.ReadUInt32LittleEndian
    : BinaryPrimitives.ReadUInt32BigEndian;
```

```csharp
// ... [SNIP]

while (size > 12)
{
    a += ReadUInt32CorrectEndian(data.Slice(index));
    b += ReadUInt32CorrectEndian(data.Slice(index + 4));
    c += ReadUInt32CorrectEndian(data.Slice(index + 8));

// ... [SNIP]
```

I did not (yet) have a chance to run another benchmark pass.

**Comment by danielmarbach** [comment]
Could we use a method that can be inlined that uses something like

``` 
BitConverter.IsLittleEndian ? BinaryPrimitives.ReverseEndianness(tempVal) : tempVal
```

?

**Comment by danielmarbach** [comment]
I can rerun the perf test later tonight with a proposal

**Comment by jsquire** [comment]
I generated a test bed when I was doing the compat runs, if that helps.     It's in a simple JSON array format.   
[partition-key-hashes.txt](https://github.com/Azure/azure-sdk-for-net/files/8352724/partition-key-hashes.txt)

**Comment by danielmarbach** [comment]
Hmmm... It does do an endian check.

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/ReaderBigEndian.cs#L127

I need to have a closer look what's going on there

**Comment by danielmarbach** [comment]
> I'm open to it, sure. I don't think that CLS compliant comes into play unless we're doing something with the public API, no?

Yeah you are right. I haven't had to worry about CLS compliance for a while... Completely forgot that part

**Comment by danielmarbach** [comment]
I think I will simply bring the span based API that netstandard 2.1 has into the lib. Then we are using the BitConverter approach

https://docs.microsoft.com/en-us/dotnet/api/system.bitconverter.toint32?view=net-6.0

**Comment by danielmarbach** [comment]
@jsquire I'm starting to wonder if this is not a bug in the original implementation. `BitConverter.ToUInt32` does an unaligned read. This means the compute hash will depending on the endianness for the same input return different output. Was that intended? Wouldn't you want for the same partition key input always the same output?

**Comment by jsquire** [comment]
> @jsquire I'm starting to wonder if this is not a bug in the original implementation. `BitConverter.ToUInt32` does an unaligned read. This means the compute hash will depending on the endianness for the same input return different output. Was that intended? 

Hrmm....  I hadn't stopped to question the reference version, but yeah, I can see where you're coming from.

> Wouldn't you want for the same partition key input always the same output?

yes, we do.    We also need this to be consistent across languages, which may be a challenge.     Let me verify whether our other SDKs can force that kind of low-level consistency.

**Comment by jsquire** [comment]
The more that I thought about this over the weekend and played around, the more I'm convinced that your observation is spot on - we should assume a consistent endian-ness for the conversion.   I think this means that your original version was right.   I would ask that we assume little endian rather than big, though, since that will be natural to the majority of hosts.

Thoughts?

**Comment by danielmarbach** [comment]
>  I would ask that we assume little endian rather than big, though, since that will be natural to the majority of hosts.

Done. Do we need to worry about backward compatibility?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs]: FYI There was recently an interesting conversation that has also been captured in 

https://github.com/dotnet/docs/issues/28823

I did run some tests and if I switch the pattern the numbers look like

image

but by adding `SkipLocalsInit` I'm now getting an impressive boost

image

**Comment by jsquire** [comment]
> Done. Do we need to worry about backward compatibility?

Thankfully, no.  We're running with a slightly different version than the service currently uses, by request, and we've set expectations that the buffered producer will assign partitions differently than the producer.   Once we GA, we'll need to avoid client-side differences, but we're still in beta.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:97]:
```suggestion
        ///
        private static short GenerateHashCode(string partitionKey)
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:154]:
```suggestion
                a += BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(index));
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:11]:
```suggestion
    /// <summary>
    ///   Encoding extensions used to provide a compatibility shim for encoding methods that are missing in .NET Standard 2.0.
    /// </summary>
    ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:24]:
```suggestion
        /// <summary>
        ///   Encodes into a span of bytes a set of characters from the specified read-only span.
        /// </summary>
        ///
        /// <param name="encoding">The encoding to be used.</param>
        /// <param name="src">The span containing the set of characters to encode.</param>
        /// <param name="dest">The byte span to hold the encoded bytes.</param>
        ///
        /// <returns>The count of encoded bytes.</returns>
        ///
        /// <remarks>
        ///   The method was introduced as a compatibility shim for .NET Standard and can be replaced should the
        ///   SDK change targets to .NET Standard 2.1 or target frameworks that provides those methods out of the box.
        ///   During  reviews it was decided to not multi-target due to the added complexity.
        ///  </remarks>
        ///
        ///   <seealso="https://docs.microsoft.com/dotnet/api/system.text.encoding.getbytes?view=netstandard-2.1#system-text-encoding-getbytes(system-readonlyspan((system-char))-system-span((system-byte)))" />
        ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EncodingExtensions.cs:61]:
```suggestion
        /// <summary>
        ///   Calculates the number of bytes produced by encoding the characters in the specified character span.
        /// </summary>
        ///
        /// <param name="encoding">The encoding to be used.</param>
        /// <param name="src">The span of characters to encode.</param>
        ///
        /// <returns>The count of bytes produced by encoding the specified character span.</returns>
        ///
       /// <remarks>
        ///   The method was introduced as a compatibility shim for .NET Standard and can be replaced should the
        ///   SDK change targets to .NET Standard 2.1 or target frameworks that provides those methods out of the box.
        ///   During  reviews it was decided to not multi-target due to the added complexity.
        ///  </remarks>
        ///
        ///   <seealso="https://docs.microsoft.com/dotnet/api/system.text.encoding.getbytes?view=netstandard-2.1#system-text-encoding-getbytes(system-readonlyspan((system-char))-system-span((system-byte)))" />
        ///
```

**Review by jsquire — APPROVED** [review]
Thank you, @danielmarbach, as always, for your taking this on and your flexibility talking through the ambiguity.  Pedantic formatting guy made one last pass, but I think the important parts are good to go.

**Comment by danielmarbach** [comment]
@jsquire let me know what you think about the recent discussion point I brought up regarding following more the best practices of always stack allocating the max size and how in that scenario having `SkipLocalsInit` would actually improve things quite a bit.

On another note: Would you be able to run your fancy script against the latest changes?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs]: Or in other words I would like to bring back https://github.com/Azure/azure-sdk-for-net/pull/27598/commits/14912a3cfda8253debf92db9944c62a0d25ffcc2

Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs]: Brought the SkipLocalsInit back and aligned the stack alloc pattern to always go for the max size as suggested by the linked conversation

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs:9]:
```suggestion
    /// <summary>
    ///   Used to indicate to the compiler that the <c>.locals init</c> flag should not be set in method headers.
    /// </summary>
    ///
    /// <remarks>Internal copy of the .NET 5 attribute.</remarks>
    ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs:4]: Do we need to wrap this in an `#ifdef` or is it just a marker interface that would have the same effect as the actual system implementation?

**Comment by jsquire** [comment]
> let me know what you think about the recent discussion point I brought up regarding following more the best practices of always stack allocating the max size and how in that scenario having `SkipLocalsInit` would actually improve things quite a bit.

I'm not opposed to bringing it back; I just question how much of an impact it will have with our `netstandard2.0` target.  If it's being evaluated by the jitter and would impact newer runtimes for applications using the SDK, I think it's a definite win.

**Comment by jsquire** [comment]
> On another note: Would you be able to run your fancy script against the latest changes?

Will do.   I'll do another pass today.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/SkipLocalsInitAttribute.cs]: Done

**Comment by danielmarbach** [comment]
> > let me know what you think about the recent discussion point I brought up regarding following more the best practices of always stack allocating the max size and how in that scenario having `SkipLocalsInit` would actually improve things quite a bit.
> 
> I'm not opposed to bringing it back; I just question how much of an impact it will have with our `netstandard2.0` target. If it's being evaluated by the jitter and would impact newer runtimes for applications using the SDK, I think it's a definite win.

It instructs the compiler to not add the locals init. So, the important part is to compile the code that uses it with a roslyn compiler version that supports that. I guess that is given here right?

**Comment by jsquire** [comment]
> It instructs the compiler to not add the locals init. So, the important part is to compile the code that uses it with a roslyn compiler version that supports that. I guess that is given here right?

I'll be interesting to see if the net6 SDK compiler does the right thing for a netstandard2.0 target.  We'll always be current with the SDK/compiler versions, it's the targets that will lag behind for compatibility.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:119]: Should we move this to a `finally` block?

**Comment by jsquire** [comment]
> > On another note: Would you be able to run your fancy script against the latest changes?
> 
> Will do. I'll do another pass today.

Test pass is solid.   260,000+ random key variations with different patterns and results are consistent.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs
Comment (danielmarbach) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:119]: Since this is perf sensitive I would lean towards leaving as is but can understand if you play the safe card 💳😊

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:119]: Talk me off the ledge.  😄    We've got user-provided input that we're assuming we can treat as UTF-8 bytes.   Failure causes a rented buffer to be orphaned.    However, I can't think of a reasonable way that you could create a normal .NET string that would cause an encoding failure.... you?

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Core/PartitionResolver.cs:119]: You and me both.   Let's leave things as-is for now.   We can always guard if we have feedback related to errors.

**Comment by jsquire** [comment]
@danielmarbach:  Looks like we're all wrapped up and ready.   Please let me know if you'd like me to merge or if you're still making tweaks.

**Comment by danielmarbach** [comment]
Go ahead

**Pull request merged**
